src/box/memtx_tx.c

```diff
@@ -3341,36 +3458,55 @@ memtx_tx_nearby_gap_item_new(struct txn *txn, enum iterator_type type,
 
 	item->type = type;
 	item->part_count = part_count;
-	const char *tmp = key;
-	for (uint32_t i = 0; i < part_count; i++)
-		mp_next(&tmp);
-	item->key_len = tmp - key;
-	if (part_count == 0) {
-		item->key = NULL;
-	} else if (item->key_len <= sizeof(item->short_key)) {
-		item->key = item->short_key;
-	} else {
-		item->key = memtx_tx_xregion_alloc(txn, item->key_len,
-						   MEMTX_TX_ALLOC_TRACKER);
-	}
-	memcpy((char *)item->key, key, item->key_len);
+	item->key = memtx_tx_save_key(txn, key, part_count, item->short_key,
+				      sizeof(item->short_key), &item->key_len);
 	return item;
 }
 
 /**
- * Allocate and create count gap item.
+ * Allocate and create count gap item. References the @a until tuple if any.
  * Note that in_read_gaps base member must be initialized later.
  */
 static struct count_gap_item *
 memtx_tx_count_gap_item_new(struct txn *txn, enum iterator_type type,
-			    const char *key, uint32_t part_count)
+			    const char *key, uint32_t part_count,
+			    struct tuple *until, hint_t until_hint)
 {
-	struct count_gap_item *item = (struct count_gap_item *)
-		memtx_tx_nearby_gap_item_new(txn, type, key, part_count);
-	item->base.base.type = GAP_COUNT;
+	assert(until == NULL || tuple_has_flag(until, TUPLE_IS_DIRTY));
+
+	struct memtx_tx_mempool *pool = &txm.count_gap_item_mempool;
+	struct count_gap_item *item = memtx_tx_xmempool_alloc(txn, pool);
+	gap_item_base_create(&item->base, GAP_COUNT, txn);
+
+	item->type = type;
+	item->part_count = part_count;
+	item->key = memtx_tx_save_key(txn, key, part_count, item->short_key,
+				      sizeof(item->short_key), &item->key_len);
+	item->until = until;
+	item->until_hint = until_hint;
+
+	/*
+	 * The @a until tuple may be invisible to the counting transaction and
+	 * may be deleted later on, let's ref it to prevent its deletion until
+	 * this gap item is deleted.
+	 */
+	if (item->until)
+		tuple_ref(item->until);
```

Comment:
This must be revised after the TX management is documented cause the tuple won't be deleted until it's story is, but the lifetime of stories is not clear enough to me yet.

Reply:
The tuple is dirty, so there's a story hilding it from the deletion. And the story won't be deleted until the last gap item that uess it is deleted, but since the count tracking is performed after the ruple had been clarified, the tuple stories has `GAP_INPLACE` items in their `read_gaps`, which will only be deleted after all transaction's gap items are removed (including this count).

So it's OK to not reference the tuple here.